substrate/network/src/on_demand.rs

```diff
@@ -0,0 +1,374 @@
+}
+
+/// On-demand requests service. Dispatches requests to appropriate peers.
+pub struct OnDemand<E: service::ExecuteInContext> {
+	core: Mutex<OnDemandCore<E>>,
+}
+
+/// On-demand response.
+pub struct Response {
+	receiver: Receiver<message::RemoteCallResponse>,
+}
+
+#[derive(Default)]
+struct OnDemandCore<E: service::ExecuteInContext> {
+	service: Weak<E>,
+	next_request_id: u64,
+	pending_requests: VecDeque<Request>,
+	active_peers: LinkedHashMap<PeerId, Request>,
+	idle_peers: VecDeque<PeerId>,
+}
+
+struct Request {
+	id: u64,
+	timestamp: Instant,
+	sender: Sender<message::RemoteCallResponse>,
+	message: message::RemoteCallRequest,
+}
+
+impl Response {
+	fn wait(&mut self) -> client::error::Result<(Vec<u8>, Vec<Vec<u8>>)> {
+		self.receiver.recv().map(|r| (r.value, r.proof)).map_err(|_| unimplemented!())
+	}
+}
+
+impl<E> OnDemand<E> where E: service::ExecuteInContext {
+	/// Creates new on-demand service.
+	pub fn new() -> Self {
+		OnDemand {
+			core: Mutex::new(OnDemandCore {
+				service: Weak::new(),
+				next_request_id: 0,
+				pending_requests: VecDeque::new(),
+				active_peers: LinkedHashMap::new(),
+				idle_peers: VecDeque::new(),
+			})
+		}
+	}
+
+	/// Sets weak reference to network service.
+	pub fn set_service_link(&self, service: Weak<E>) {
+		self.core.lock().service = service;
+	}
+
+	/// Execute method call on remote node, returning execution result and proof.
+	pub fn remote_call(&self, block: HeaderHash, method: &str, data: &[u8]) -> Response {
+		let (sender, receiver) = channel();
+		let result = Response {
+			receiver: receiver,
+		};
+
+		{
+			let mut core = self.core.lock();
+			core.insert(sender, message::RemoteCallRequest {
+				id: 0,
+				block: block,
+				method: method.into(),
+				data: data.to_vec(),
+			});
+			core.dispatch();
+		}
+
+		result
+	}
+}
+
+impl<E> OnDemandService for OnDemand<E> where E: service::ExecuteInContext {
+	fn on_connect(&self, peer: PeerId, role: service::Role) {
+		if !role.intersects(service::Role::FULL | service::Role::COLLATOR | service::Role::VALIDATOR) { // TODO: correct?
+			return;
+		}
+
+		let mut core = self.core.lock();
+		core.add_peer(peer);
+		core.dispatch();
+	}
+
+	fn on_disconnect(&self, peer: PeerId) {
+		let mut core = self.core.lock();
+		core.remove_peer(peer);
+		core.dispatch();
+	}
+
+	fn maintain_peers(&self, io: &mut SyncIo) {
+		let mut core = self.core.lock();
+		for bad_peer in core.maintain_peers() {
+			trace!(target: "sync", "Remote request timeout for peer {}", bad_peer);
+			io.disconnect_peer(bad_peer);
+		}
+		core.dispatch();
+	}
+
+	fn on_remote_response(&self, io: &mut SyncIo, peer: PeerId, response: message::RemoteCallResponse) {
+		let mut core = self.core.lock();
+		match core.remove(peer, response.id) {
+			Some(request) => {
+				// we do not bother if receiver has been dropped already
+				let _ = request.sender.send(response);
```

Comment:
it should check the validity of the response here and reassign if the response is not correct.

Reply:
we could have each request come with an associated `attempts` counter, and users can specify a default maximum when creating the OnDemand service. When dispatching a request they can use that default or provide their own maximum.

Reply:
Maybe better to add this logic to the `LightCallExecutor` / `LightBackend` - remote execution proof check = local execution (am I right?). And that is performed in `LightCallExecutor` right now. So fetch => execute locally => disconnect if fail.

Reply:
sure, as long as there's a convenient way to disconnect the peer from the `LightCallExecutor`

Reply:
This is a little different than the OnDemand service in parity-ethereum, which handles data validation and rescheduling internally. But then it has the nice property that only good data comes out of the service. This architecture is also useful for implementing a network protocol that implements heterogeneous requests in a single packet, where inputs of later requests are outputs of earlier ones (see PIP).

Reply:
(we can defer that to another PR, but it might be useful to think of the future plan when continuing)

Reply:
no - I think this would change the **structure** significantly, so let it be in this PR - I'll update the code here

Reply:
I've added `FetchChecker` trait to the light client module (where `Fetcher` is also defined). It is responsible for checking remote response and converting it from 'remote' to 'local' form (like dropping merkle proofs, because they're not required after check is performed).
So implementation of `Fetcher` (i.e. `OnDemand`) could stay in network crate && do not refer directly to blockchain/client && only check+convert remote response using this trait (implemented in `light` module).

Reply:
I don't see `FetchChecker` anywhere in the diff.

Reply:
Trait and implementation: https://github.com/paritytech/polkadot/pull/150/commits/f4070e2b1732353112b3a08382a7f1f05817323a#diff-389a8eabc40f446e244b84b3d9c5c343R51
Usage: https://github.com/paritytech/polkadot/pull/150/commits/f4070e2b1732353112b3a08382a7f1f05817323a#diff-a15b054a65ff0e7c42d5add3417c90faR151